### PR TITLE
Preliminary game core

### LINKED ISSUE #1: init dev env
1. import https://github.com/SailGame/proto   (use submodule)
2. import gtest/gmock,   grpc (submodule or fetchcontent or whatever it is)
3. set up git action (optional)
4. set up clang-format
5. set up CMakeList skeleton.

### PR DESCRIPTION
**Pull request by MrHate**
Still working on the game core. Insurance of an enough complete core decreases unnecessary refactoring.
**Do not merge presently.**

### CODE REVIEW

File: TexasScore.cpp [hunk 1]
```diff
@@ -1,28 +1,35 @@
...
+int high_card(const cardarr_t &cards);
 
-  // 1. Royal Straight Flush
+int _compare(const int a, const int b);
+int _for_each_combination(std::function<int(const cardarr_t)> &lambda,
+                          const cardarr_t &cards);
+} // namespace
 
+const Dummy::score_t poker::_score(cardarr_t cards) {
+  std::sort(cards.begin(), cards.end(), [](int a, int b) { return a > b; });
```
Comment (Veiasai) [inline TexasScore.cpp]: `std::sort(begin, end, std::greater<int>())`

File: TexasScore.cpp [hunk 2]
```diff
@@ -32,24 +39,203 @@ const int Dummy::Score(const uid_t uid) {
...
+const Dummy::score_t Dummy::Score(const uid_t uid) {
+  // TODO: Record the winner cards into attribute?
+  auto cards = holecards[uid];
+  cards.insert(cards.end(), board.begin(), board.end());
+  return poker::_score(cards);
+}
+
+int CardScore::Compare(const CardScore &rhs) const {
+  int ret;
+  if ((ret = _compare(royal_straight_flush, rhs.royal_straight_flush)))
```
Comment (Veiasai) [inline TexasScore.cpp]: this is c-style, not c++ 17.

it should be if(ret = _compare(); ret)

File: Dummy.cpp [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+  }
+  roundbets[uid] = bet;
+
+  if (uid == cutoff) {
+    if (raised) {
+      // Someone raised within this turn.
+      raised = false;
+    } else {
+      // All checked.
+      if (board.size() < 3) {
```
Comment (Veiasai) [inline Dummy.cpp]: use named variable 3, 5

File: Dummy.cpp [hunk 4]
```diff
@@ -0,0 +1,201 @@
...
+    return INVALID_BET;
+
+  if (bet > cur_chips) {
+    // Raise
+    raised = true;
+    cur_chips = bet;
+  }
+  roundbets[uid] = bet;
+
+  if (uid == cutoff) {
```
Comment (Veiasai) [inline Dummy.cpp]: what is cutoff

File: Dummy.h [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+  const status_t Begin();
+  const uid_t Join(const std::string &addr);
+  // positive -> raise or call; zero -> check; -1 -> conceed.
+  const status_t Play(const uid_t uid, const chip_t value);
+
+  const GameStatus DumpStatusForUser(const uid_t uid) const;
+  const void DumpDebugMessages(std::ostream &out);
+
+  const uid_t PreviousWinner() const { return prev_winner; }
+  const uid_t FirstPlayer() const { return 1; }
```
Comment (Veiasai) [inline Dummy.h:51]: should the firstplayer and lastplayer consider someone all-in or fold?
say if the uid == 1 is all-in,  we should skip the execution? do you handle this case?

**Review by Veiasai — CHANGES_REQUESTED** [review]

File: Dummy.h [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+
+  int Compare(const CardScore &rhs) const;
+};
+
+namespace poker {
+
+// clang-format off
+  enum {
+    H1 = 0x11, H2, H3, H4, H5, H6, H7, H8, H9, H10, HJ, HQ, HK, // Hearts
+    S1 = 0x21, S2, S3, S4, S5, S6, S7, S8, S9, S10, SJ, SQ, SK, // Spides
```
Comment (Gusabary) [inline Dummy.h]: typo, should be Spades

File: Dummy.h [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+
+class Dummy {
+public:
+  using uid_t = int;
+  using chip_t = int;
+  using card_t = int;
+  using status_t = int;
+  using score_t = CardScore;
+
+  enum { STOP = -1, READY = 0, NOT_YOUR_TURN = 1, INVALID_BET = 2 };
```
Comment (Gusabary) [inline Dummy.h]: better to use scoped enums

File: TexasScore.cpp [hunk 8]
```diff
@@ -0,0 +1,241 @@
...
+int full_house(const cardarr_t &cards);
+int flush(const cardarr_t &cards);
+int straight(const cardarr_t &cards);
+int three_of_a_kind(const cardarr_t &cards);
+int two_pair(const cardarr_t &cards);
+int one_pair(const cardarr_t &cards);
+int high_card(const cardarr_t &cards);
+
+int _compare(const int a, const int b);
+int _for_each_combination(std::function<int(const cardarr_t)> &lambda,
```
Comment (Gusabary) [inline TexasScore.cpp]: warning: `_for_each_combination` never used

**Review by Gusabary — CHANGES_REQUESTED** [review]

File: Dummy.h (hunk 5 above)
Comment (MrHate) [reply Dummy.h:51]: that's still in todo-list.
working on score logic currently.

File: Dummy.cpp (hunk 4 above)
Comment (MrHate) [reply Dummy.cpp]: should be small blind

**Review by Veiasai — APPROVED** [review]

**Review by Gusabary — APPROVED** [review]

**Pull request merged**